### PR TITLE
Leverage ember-template-recast

### PR DESCRIPTION
**Pull request by dcyriller**
This PR proposes to use ember-template-recast to switch to a non-destructive parser. The goal is to eventually implement the `--fix` option. It would be the first step of #760.

#### four small comments:
- [merged ✅] I've extracted 3 commits into a separate PR that may be merged first: https://github.com/ember-template-lint/ember-template-lint/pull/872
- [resolved ✅] `template-recast`'s `transform` method seems to take a function that returns a visitor (not a plugin) as a second argument. Whereas the README states the opposite (it should return a plugin not a visitor). Should I make an adjustement in `template-recast`? Or is it fine to go on with a visitor as it is in this PR? (see https://github.com/ember-template-lint/ember-template-recast/pull/158)
- I need to adjust three rules to make them compatible with `template-recast` (one test failing each): `block-indentation` (✅), `no-inline-styles` (✅ with a 🙊-patch) and `linebreak-style` (✅)
- this PR relies on https://github.com/ember-template-lint/ember-template-recast/pull/154 that will need to be resolved first ✅

### CODE REVIEW

**Comment by dcyriller** [comment]
Hopefully it should be good for review now :)

**Comment by rwjblue** [comment]
A few notes here:

* Ready for a rebase after #892
* I think we will need #850 to land first (`ember-template-recast` uses `codemod` mode, and we have rules that need adjustment before we are ready). Some of the commits here seem to replicate some of those changes, but I'd rather land #850 (@initram did a bunch of work on it over there).

**Comment by dcyriller** [comment]
I didn’t see that #850 had to solve similar issues in the tests 😅. I’ll wait for it to be merged and rebase after!

**Comment by rwjblue** [comment]
Ok, just landed #850

**Comment by dcyriller** [comment]
@rwjblue it should be ready for review now 🙂

File: test/unit/rules/lint-no-inline-styles-test.js [hunk 1]
```diff
@@ -36,7 +36,20 @@ generateRuleTests({
     'style={{foo}}',
     'style="{{foo}} {{bar}}"',
     'style',
-    'style=""',
     'style="color:blue;margin-left:30px;"',
-  ].map(mapBadStyleExampleToTestRule),
+  ]
+    .map(mapBadStyleExampleToTestRule)
```
Comment (dcyriller) [inline test/unit/rules/lint-no-inline-styles-test.js]: You might want to see https://github.com/ember-template-lint/ember-template-lint/pull/873/commits/c4e29f6360d60e5f6fd9d63fcaad8d01dc1acb88 commit message

File: lib/index.js [hunk 2]
```diff
@@ -163,7 +155,7 @@ class Linter {
 
     for (let rule of rules) {
       try {
-        traverse(templateAST, buildPlugin(rule));
+        transform(templateAST, () => rule.getVisitor());
```
Comment (rwjblue) [inline lib/index.js]: Ya, wasn’t sure if you’d still want / need `buildPlugin` (now that `rule.templateEnvironmentData` isn’t used). Glad to see this simplified. :)

**Comment by rwjblue** [comment]
Even though CI passes for Node 6 (not 100% sure why, ember-template-recast must not use Node 8 specific syntax?), I still think we should block landing this on bumping majors to drop Node 6.

**Comment by dcyriller** [comment]
> Even though CI passes for Node 6 (not 100% sure why, ember-template-recast must not use Node 8 specific syntax?), I still think we should block landing this on bumping majors to drop Node 6.

Indeed, it should have failed on the step `yarn install` since `template-recast`‘s engines field states: `8 || 10 || >=12` (see https://github.com/ember-template-lint/ember-template-recast/blob/v3.2.8/package.json#L56).

The reason why it doesn’t fail is because this yarn flag in the `travis.yml` file: `—ignore-engines` (see https://github.com/ember-template-lint/ember-template-lint/blob/v1.7.0/.travis.yml#L19). I guess it is there to be able to cache the install step between various CI jobs (testing against different node versions).

So, this PR is definitely blocked by a Node requirement upgrade. :)

**Comment by rwjblue** [comment]
FWIW, I added `—ignore-engines` so that I can use `release-it` (it requires Node 8).

**Comment by dcyriller** [comment]
I now understand your previous comment 🙈

**Comment by rwjblue** [comment]
Rebased and targeted `next` branch (follow along in #897).

**Pull request merged**